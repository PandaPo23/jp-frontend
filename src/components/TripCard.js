import React, { useMemo } from 'react';
import { toJS } from 'mobx';
import { inject, observer } from 'mobx-react';
import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import compose from 'lodash/fp/compose';
import fp from 'lodash/fp';
import PropTypes from 'prop-types';

import {
  Box,
  Card,
  Divider,
  Flex,
  FlexPanel,
  Heading5,
  ImageSorter,
  Text,
} from '../design-system';
import { durationToDays } from '../utils/date';
import { TravelStylesDistributionView } from './';

const shuffleArray = (arr) => arr.sort(() => Math.random() - 0.5);

const getAllImagesFromTrip = (trip) =>
  fp.compose(
    shuffleArray,
    fp.uniqBy('src'),
    fp.reduce(
      (images, destination) => [...images, ...toJS(destination.images || [])],
      toJS(trip.images) || []
    )
  )(trip.destinations || []);

export const TripCard = ({
  trip,
  t,
  setCurrentTrip,
  travelStyleLabels,
  history,
  mapIsInitialized,
  prefAccommodation,
  accommodationsOptions,
  ...cardProps
}) => {
  const tripImages = useMemo(() => getAllImagesFromTrip(trip), [trip]);
  const accommodationItem = accommodationsOptions.find(
    (item) => item.value === prefAccommodation
  );
  return (
    <Card {...cardProps}>
      <Flex>
        <Flex flex={1.5} height="auto">
          {mapIsInitialized && (
            <ImageSorter
              borderRadius="left.sm"
              images={tripImages}
              height="100%"
            />
          )}
        </Flex>
        <Box
          p={3}
          flex={2}
          hoverable
          onClick={() => history.push(`/trips/${trip.id}`)}
        >
          <FlexPanel>
            <FlexPanel justifyContent="space-around">
              <Heading5 m={0} textStyle="capitalize">
                {trip.name}
              </Heading5>
              <Text mt={1} mb={3} fontSize={3} uppercase>
                {t('list_format', {
                  list: trip.destinations.map((d) => d.name),
                })}
              </Text>
            </FlexPanel>
            <Flex justifyContent="space-between">
              <FlexPanel flex={1} alignItems="center">
                <Text fontSize={4} bold>
                  {t('dollar_amount', {
                    amount:
                      trip.accommodation[prefAccommodation] ||
                      trip.estimatedCost ||
                      'N/A',
                  })}
                </Text>
                <Text
                  pt={2}
                  fontSize={2}
                  color="misc.muted"
                  uppercase
                  textAlign="center"
                >
                  {t('per_person', 'per person')}
                </Text>
                <Text fontSize={4} bold mt={2}>
                  {accommodationItem && accommodationItem.label}
                </Text>
              </FlexPanel>
              <Divider vertical height={100} />
              <FlexPanel flex={1} alignItems="center">
                <Text fontSize={4} bold>
                  {durationToDays(trip.duration)}
                </Text>
                <Text pt={2} fontSize={2} uppercase color="misc.muted" center>
                  {t('day', { count: durationToDays(trip.duration) })}
                </Text>
              </FlexPanel>
              <Divider vertical height={100} />
              <Flex fontSize={2} justifyContent="center" flex={2}>
                <TravelStylesDistributionView values={trip.travelStyles} />
              </Flex>
            </Flex>
          </FlexPanel>
        </Box>
      </Flex>
    </Card>
  );
};

TripCard.propTypes = {
  trip: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
};

const TripCardWrapper = (props) => <TripCard {...props} />;

export default compose(
  withRouter,
  withTranslation('common'),
  inject(({ app }, { trip }) => ({
    mapIsInitialized: app.map.initialized,
    accommodationsOptions: app.preferences.accommodationsOptions || [],
    prefAccommodation: app.customer.getTripAccommodation(trip.id),
  })),
  observer
)(TripCardWrapper);
