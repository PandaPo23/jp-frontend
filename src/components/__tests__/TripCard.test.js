import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { TripCard } from '../TripCard';

jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
  Map: () => ({}),
}));

const mockMath = Object.create(global.Math);
mockMath.random = () => 0.5;
global.Math = mockMath;

const images = [
  {
    caption: 'San Cristobal De Las Casas',
    src:
      'https://jubelproduction.s3.amazonaws.com/uploads/destination/main_image/547/shutterstock_170168309.jpg',
    thumbnail:
      'https://jubelproduction.s3.amazonaws.com/uploads/destination/main_image/547/card_shutterstock_170168309.jpg',
  },
  {
    caption: 'San Cristobal De Las Casas',
    src:
      'https://jubelproduction.s3.amazonaws.com/uploads/photo/image/269/shutterstock_370346327.jpg',
    thumbnail:
      'https://jubelproduction.s3.amazonaws.com/uploads/photo/image/269/thumb_shutterstock_370346327.jpg',
  },
  {
    caption: 'San Cristobal De Las Casas',
    src:
      'https://jubelproduction.s3.amazonaws.com/uploads/photo/image/270/shutterstock_518526283.jpg',
    thumbnail:
      'https://jubelproduction.s3.amazonaws.com/uploads/photo/image/270/thumb_shutterstock_518526283.jpg',
  },
];

const itinerary = {
  name: 'Trip Name Goes Here',
  destinations: ['Seulla', 'Cadiz'],
  price: 1184,
  startDate: 'Jun 20, 2019',
  duration: 'P17D',
  numberOfDays: 12,
  images,
  accommodation: {
    budget: 2352,
    hostel: 1176,
    luxe: 6114,
    mid_range: 3762,
    ultra: 11758,
  },
  travelStyle: {
    oc: 20,
    an: 34,
    pp: 7,
    cu: 45,
    cn: 93,
  },
};

const accommodationsOptions = [
  { value: 'Hostel', label: 'Hostel' },
  { value: 'Budget', label: 'Budget' },
  { value: 'Mid-Range', label: 'Mid-Range' },
  { value: 'Luxe', label: 'Luxe' },
  { value: 'Ultra', label: 'Ultra' },
];

describe('TripCard component', () => {
  it('renders correctly', () => {
    const localProps = {
      trip: itinerary,
      accommodationsOptions,
      accommodation: 'Mid-Range',
      t: (v) => v,
    };
    const wrapper = shallow(<TripCard {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
