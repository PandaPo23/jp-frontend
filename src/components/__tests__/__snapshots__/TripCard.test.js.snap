// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripCard component renders correctly 1`] = `
<Card
  accommodation="Mid-Range"
  bg="surface"
  borderRadius="sm"
  boxShadow={4}
  className="surface"
>
  <Flex
    display="flex"
  >
    <Flex
      display="flex"
      flex={1.5}
      height="auto"
    />
    <Box
      flex={2}
      hoverable={true}
      onClick={[Function]}
      p={3}
    >
      <FlexPanel
        display="flex"
        flexDirection="column"
      >
        <FlexPanel
          display="flex"
          flexDirection="column"
          justifyContent="space-around"
        >
          <Heading5
            as="h5"
            className="heading"
            m={0}
            textStyle="capitalize"
          >
            Trip Name Goes Here
          </Heading5>
          <Text
            fontSize={3}
            mb={3}
            mt={1}
            uppercase={true}
          >
            list_format
          </Text>
        </FlexPanel>
        <Flex
          display="flex"
          justifyContent="space-between"
        >
          <FlexPanel
            alignItems="center"
            display="flex"
            flex={1}
            flexDirection="column"
          >
            <Text
              bold={true}
              fontSize={4}
            >
              dollar_amount
            </Text>
            <Text
              color="misc.muted"
              fontSize={2}
              pt={2}
              textAlign="center"
              uppercase={true}
            >
              per_person
            </Text>
            <Text
              bold={true}
              fontSize={4}
              mt={2}
            />
          </FlexPanel>
          <Divider
            height={100}
            vertical={true}
          />
          <FlexPanel
            alignItems="center"
            display="flex"
            flex={1}
            flexDirection="column"
          >
            <Text
              bold={true}
              fontSize={4}
            >
              17
            </Text>
            <Text
              center={true}
              color="misc.muted"
              fontSize={2}
              pt={2}
              uppercase={true}
            >
              day
            </Text>
          </FlexPanel>
          <Divider
            height={100}
            vertical={true}
          />
          <Flex
            display="flex"
            flex={2}
            fontSize={2}
            justifyContent="center"
          >
            <withI18nextTranslation(TravelStylesDistributionView) />
          </Flex>
        </Flex>
      </FlexPanel>
    </Box>
  </Flex>
</Card>
`;
